File: ltms/go/ltms_ex.go
// LTMS examples and tests.
// Converted from ltms-ex.lisp.
package ltms

import "fmt"

// TestExplain tests the explain functionality.
func TestExplain() {
	l := CreateLTMS("Explain Example", LTMSOptions{
		NodeString:             DefaultNodeString,
		CheckingContradictions: true,
		ContradictionHandler:   AskUserHandler,
		CacheDatums:            true,
		DelaySat:               true,
	})
	TmsCreateNode(l, "x", true)
	NormLTMS = l
	clauses := Normalize(l, []interface{}{":OR", "x", "y"})
	for _, clause := range clauses {
		s := SimplifyClause(clause)
		if s != nil {
			AddClauseInternal(s, nil, true)
		}
	}
	clauses = Normalize(l, []interface{}{":OR", []interface{}{":NOT", "y"}, "z"})
	for _, clause := range clauses {
		s := SimplifyClause(clause)
		if s != nil {
			AddClauseInternal(s, nil, true)
		}
	}
	clauses = Normalize(l, []interface{}{":OR", []interface{}{":NOT", "z"}, "r"})
	for _, clause := range clauses {
		s := SimplifyClause(clause)
		if s != nil {
			AddClauseInternal(s, nil, true)
		}
	}
	EnableAssumption(FindNode(l, "x"), LabelFalse)
	ExplainNode(FindNode(l, "r"))
}

// TestFormula tests formula normalization.
func TestFormula(complete bool) *LTMS {
	var cm CompleteMode
	if complete {
		cm = CompleteTrue
	}
	l := CreateLTMS("Formula", LTMSOptions{
		NodeString:             DefaultNodeString,
		CheckingContradictions: true,
		ContradictionHandler:   AskUserHandler,
		CacheDatums:            true,
		Complete:               cm,
		DelaySat:               true,
	})
	if complete {
		InitCompleteLTMS(l)
	}
	r := TmsCreateNode(l, "r", false)
	s := TmsCreateNode(l, "s", false)
	tt := TmsCreateNode(l, "t", false)
	u := TmsCreateNode(l, "u", false)
	_ = r
	_ = s
	_ = tt
	_ = u
	AddFormula(l, []interface{}{":IMPLIES",
		[]interface{}{":AND", r, []interface{}{":IMPLIES", s, tt}}, u}, nil)
	return l
}

// TestAsk tests the ask/user interface.
func TestAsk() {
	l := CreateLTMS("Testing asking", LTMSOptions{
		NodeString:             DefaultNodeString,
		CheckingContradictions: true,
		ContradictionHandler:   AskUserHandler,
		CacheDatums:            true,
		DelaySat:               true,
	})
	n1 := TmsCreateNode(l, "N1", true)
	n2 := TmsCreateNode(l, "N2", true)
	EnableAssumption(n1, LabelFalse)
	EnableAssumption(n2, LabelFalse)
	AddFormula(l, []interface{}{":OR", n1, n2}, nil)
	WhyNodes(l)
}

// TestAvoidAll tests the avoid-all contradiction handler.
func TestAvoidAll() {
	l := CreateLTMS("Testing avoid all", LTMSOptions{
		NodeString:             DefaultNodeString,
		CheckingContradictions: true,
		ContradictionHandler:   AvoidAll,
		CacheDatums:            true,
		DelaySat:               true,
	})
	n1 := TmsCreateNode(l, "N1", true)
	n2 := TmsCreateNode(l, "N2", true)
	EnableAssumption(n1, LabelFalse)
	EnableAssumption(n2, LabelFalse)
	AddFormula(l, []interface{}{":OR", n1, n2}, nil)
	WhyNodes(l)
}

// Test1 tests basic resolution.
func Test1(complete bool) {
	var cm CompleteMode
	if complete {
		cm = CompleteTrue
	}
	l := CreateLTMS("TEST1", LTMSOptions{
		NodeString:             DefaultNodeString,
		CheckingContradictions: true,
		ContradictionHandler:   AskUserHandler,
		CacheDatums:            true,
		Complete:               cm,
		DelaySat:               true,
	})
	if complete {
		InitCompleteLTMS(l)
	}
	x := TmsCreateNode(l, "x", false)
	y := TmsCreateNode(l, "y", false)
	_ = y
	AddFormula(l, []interface{}{":OR", x, y}, nil)
	AddFormula(l, []interface{}{":OR", x, []interface{}{":NOT", y}}, nil)
	CompleteLTMS(l)
	if !TrueNode(x) {
		panic("TEST1 failed")
	}
	fmt.Println("TEST1 passed")
}

// TestBug tests a known LTMS bug scenario.
func TestBug() {
	l := CreateLTMS("BUG check", LTMSOptions{
		NodeString:             DefaultNodeString,
		CheckingContradictions: true,
		ContradictionHandler:   AskUserHandler,
		CacheDatums:            true,
		DelaySat:               true,
	})
	x := TmsCreateNode(l, "x", true)
	y := TmsCreateNode(l, "y", true)
	z := TmsCreateNode(l, "z", false)
	_ = z
	AddFormula(l, []interface{}{":OR", x, z}, nil)
	AddFormula(l, []interface{}{":OR", y, z}, nil)
	EnableAssumption(x, LabelFalse)
	EnableAssumption(y, LabelFalse)
	WhyNodes(l)
	RetractAssumption(x)
	WhyNodes(l)
}

// TestBug1 tests bug with complete LTMS.
func TestBug1(complete bool) {
	var cm CompleteMode
	if complete {
		cm = CompleteTrue
	}
	l := CreateLTMS("BUG check", LTMSOptions{
		NodeString:             DefaultNodeString,
		CheckingContradictions: true,
		ContradictionHandler:   AskUserHandler,
		CacheDatums:            true,
		Complete:               cm,
		DelaySat:               true,
	})
	if complete {
		InitCompleteLTMS(l)
	}
	x := TmsCreateNode(l, "x", true)
	y := TmsCreateNode(l, "y", true)
	z := TmsCreateNode(l, "z", false)
	_ = z
	AddFormula(l, []interface{}{":OR", x, z}, nil)
	AddFormula(l, []interface{}{":OR", y, z}, nil)
	EnableAssumption(x, LabelFalse)
	EnableAssumption(y, LabelFalse)
	WhyNodes(l)
	RetractAssumption(x)
	WhyNodes(l)
}

// TestTax tests taxonomy constraints.
func TestTax(n int, complete bool) {
	var cm CompleteMode
	if complete {
		cm = CompleteTrue
	}
	l := CreateLTMS("taxing", LTMSOptions{
		NodeString:             DefaultNodeString,
		CheckingContradictions: true,
		ContradictionHandler:   AskUserHandler,
		CacheDatums:            true,
		Complete:               cm,
		DelaySat:               true,
	})
	if complete {
		InitCompleteLTMS(l)
	}
	var nodes []*TmsNode
	for i := 0; i < n; i++ {
		nodes = append(nodes, TmsCreateNode(l, i, false))
	}
	taxForm := make([]interface{}, 0, n+1)
	taxForm = append(taxForm, ":TAXONOMY")
	for _, node := range nodes {
		taxForm = append(taxForm, node)
	}
	AddFormula(l, taxForm, nil)
	fmt.Printf("\n %d prime implicates", l.ClauseCounter)
}

// TestE tests with assumption nodes.
func TestE(complete bool) {
	var cm CompleteMode
	if complete {
		cm = CompleteTrue
	}
	l := CreateLTMS("example", LTMSOptions{
		NodeString:             DefaultNodeString,
		Debugging:              true,
		CheckingContradictions: true,
		ContradictionHandler:   AskUserHandler,
		CacheDatums:            true,
		Complete:               cm,
		DelaySat:               true,
	})
	if complete {
		InitCompleteLTMS(l)
	}
	a := TmsCreateNode(l, "a", true)
	b := TmsCreateNode(l, "b", true)
	c := TmsCreateNode(l, "c", true)
	d := TmsCreateNode(l, "d", true)
	e := TmsCreateNode(l, "e", true)
	AddFormula(l, []interface{}{":OR", []interface{}{":NOT", a}, b}, nil)
	AddFormula(l, []interface{}{":OR", []interface{}{":NOT", c}, d}, nil)
	AddFormula(l, []interface{}{":OR", []interface{}{":NOT", c}, e}, nil)
	AddFormula(l, []interface{}{":OR", []interface{}{":NOT", b},
		[]interface{}{":NOT", d}, []interface{}{":NOT", e}}, nil)
}

// RunTests runs all basic LTMS tests.
func RunTests() {
	fmt.Println("Running TestAsk...")
	TestAsk()
	fmt.Println("\nRunning TestAvoidAll...")
	TestAvoidAll()
	fmt.Println("\nRunning TestBug...")
	TestBug()
	fmt.Println("\nRunning TestBug1...")
	TestBug1(true)
	fmt.Println("\nRunning TestTax(3)...")
	TestTax(3, true)
	fmt.Println("\nAll tests completed.")
}

File: ltms/cpp/ltms_ex.cpp
/// LTMS Examples and Tests - Implementation
/// Converted from ltms-ex.lisp
/// Copyright (c) 1986-1990, Kenneth D. Forbus, University of Illinois,
/// Johan de Kleer and Xerox Corporation. All rights reserved.

#include "ltms_ex.h"
#include <iostream>
#include <stdexcept>
#include <vector>
#include <string>

// Helper to build formula from TmsNode* values
static std::any N(TmsNode* n) { return std::any(n); }
static std::any S(const std::string& s) { return std::any(s); }

static std::any make_or(std::initializer_list<std::any> args) {
    std::vector<std::any> v;
    v.push_back(S(":OR"));
    for (auto& a : args) v.push_back(a);
    return std::any(v);
}

static std::any make_not(const std::any& arg) {
    std::vector<std::any> v;
    v.push_back(S(":NOT"));
    v.push_back(arg);
    return std::any(v);
}

static std::any make_implies(const std::any& a, const std::any& b) {
    std::vector<std::any> v;
    v.push_back(S(":IMPLIES"));
    v.push_back(a);
    v.push_back(b);
    return std::any(v);
}

static std::any make_and(std::initializer_list<std::any> args) {
    std::vector<std::any> v;
    v.push_back(S(":AND"));
    for (auto& a : args) v.push_back(a);
    return std::any(v);
}

static std::any make_tax(std::initializer_list<std::any> args) {
    std::vector<std::any> v;
    v.push_back(S(":TAXONOMY"));
    for (auto& a : args) v.push_back(a);
    return std::any(v);
}

static std::any make_iff(const std::any& a, const std::any& b) {
    std::vector<std::any> v;
    v.push_back(S(":IFF"));
    v.push_back(a);
    v.push_back(b);
    return std::any(v);
}

// ================================================================
// test_explain
// ================================================================

void test_explain() {
    auto* ltms = create_ltms("Explain Example");
    auto* x = tms_create_node(ltms, "x", true);

    // (:OR "x" "y")
    add_formula(ltms, make_or({S("x"), S("y")}));

    // (:OR (:NOT "y") "z")
    add_formula(ltms, make_or({make_not(S("y")), S("z")}));

    // (:OR (:NOT "z") "r")
    add_formula(ltms, make_or({make_not(S("z")), S("r")}));

    enable_assumption(x, NodeLabel::FALSE);
    explain_node(find_node(ltms, S("r")));
}

// ================================================================
// test_formula
// ================================================================

void test_formula(bool complete) {
    std::any comp = complete ? std::any(true) : std::any();
    auto* ltms = create_ltms("Formula", nullptr, false, true,
                              nullptr, nullptr, true, comp);
    auto* r = tms_create_node(ltms, "r");
    auto* s = tms_create_node(ltms, "s");
    auto* tt = tms_create_node(ltms, "t");
    auto* u = tms_create_node(ltms, "u");

    // (:implies (:and r (:implies s t)) u)
    add_formula(ltms, make_implies(
        make_and({N(r), make_implies(N(s), N(tt))}),
        N(u)));
}

// ================================================================
// run_ltms_tests
// ================================================================

void run_ltms_tests() {
    std::cout << "Running LTMS tests..." << std::endl;

    test_ask();
    std::cout << "\ntest_ask passed." << std::endl;

    test_avoid_all();
    std::cout << "test_avoid_all passed." << std::endl;

    test_bug();
    std::cout << "test_bug passed." << std::endl;

    test_bug1();
    std::cout << "test_bug1 passed." << std::endl;

    test_tax(3);
    std::cout << "\ntest_tax passed." << std::endl;

    test_tax1(3);
    std::cout << "\ntest_tax1 passed." << std::endl;
}

// ================================================================
// test_ask
// ================================================================

void test_ask() {
    auto* ltms = create_ltms("Testing asking");
    auto* n1 = tms_create_node(ltms, "N1", true);
    auto* n2 = tms_create_node(ltms, "N2", true);
    enable_assumption(n1, NodeLabel::FALSE);
    enable_assumption(n2, NodeLabel::FALSE);

    add_formula(ltms, make_or({N(n1), N(n2)}));
    why_nodes(ltms);
}

// ================================================================
// test_avoid_all
// ================================================================

void test_avoid_all() {
    auto* ltms = create_ltms("Testing avoid all", nullptr, false, true,
                              avoid_all);
    auto* n1 = tms_create_node(ltms, "N1", true);
    auto* n2 = tms_create_node(ltms, "N2", true);
    enable_assumption(n1, NodeLabel::FALSE);
    enable_assumption(n2, NodeLabel::FALSE);

    add_formula(ltms, make_or({N(n1), N(n2)}));
    why_nodes(ltms);
}

// ================================================================
// test1
// ================================================================

void test1(bool complete) {
    std::any comp = complete ? std::any(true) : std::any();
    auto* ltms = create_ltms("TEST1", nullptr, false, true,
                              nullptr, nullptr, true, comp);
    auto* x = tms_create_node(ltms, "x");
    auto* y = tms_create_node(ltms, "y");

    add_formula(ltms, make_or({N(x), N(y)}));
    add_formula(ltms, make_or({N(x), make_not(N(y))}));
    complete_ltms(ltms);

    if (!true_node(x)) {
        throw std::runtime_error("TEST1 failed");
    }
}

// ================================================================
// test_bug
// ================================================================

void test_bug() {
    auto* ltms = create_ltms("BUG check");
    auto* x = tms_create_node(ltms, "x", true);
    auto* y = tms_create_node(ltms, "y", true);
    auto* z = tms_create_node(ltms, "z");

    add_formula(ltms, make_or({N(x), N(z)}));
    add_formula(ltms, make_or({N(y), N(z)}));

    enable_assumption(x, NodeLabel::FALSE);
    enable_assumption(y, NodeLabel::FALSE);
    why_nodes(ltms);

    retract_assumption(x);
    why_nodes(ltms);
}

// ================================================================
// test_bug1
// ================================================================

void test_bug1(bool complete) {
    std::any comp = complete ? std::any(true) : std::any();
    auto* ltms = create_ltms("BUG check", nullptr, false, true,
                              nullptr, nullptr, true, comp);
    auto* x = tms_create_node(ltms, "x", true);
    auto* y = tms_create_node(ltms, "y", true);
    auto* z = tms_create_node(ltms, "z");

    add_formula(ltms, make_or({N(x), N(z)}));
    add_formula(ltms, make_or({N(y), N(z)}));

    enable_assumption(x, NodeLabel::FALSE);
    enable_assumption(y, NodeLabel::FALSE);
    why_nodes(ltms);

    retract_assumption(x);
    why_nodes(ltms);
}

// ================================================================
// test_tax
// ================================================================

void test_tax(int n, bool complete) {
    std::any comp = complete ? std::any(true) : std::any();
    auto* ltms = create_ltms("taxing", nullptr, false, true,
                              nullptr, nullptr, true, comp);

    std::vector<std::any> tax_args;
    tax_args.push_back(S(":TAXONOMY"));
    for (int i = 0; i < n; i++) {
        tax_args.push_back(N(tms_create_node(ltms, std::to_string(i))));
    }
    add_formula(ltms, std::any(tax_args));
    std::cout << "\n " << ltms->clause_counter << " prime implicates";
}

void test_tax1(int n) {
    auto* ltms = create_ltms("taxing");

    std::vector<std::any> tax_args;
    tax_args.push_back(S(":TAXONOMY"));
    for (int i = 0; i < n; i++) {
        tax_args.push_back(N(tms_create_node(ltms, std::to_string(i))));
    }
    add_formula(ltms, std::any(tax_args));
    std::cout << "\n " << ltms->clause_counter << " prime implicates";
}

// ================================================================
// test_e
// ================================================================

void test_e(bool complete) {
    std::any comp = complete ? std::any(true) : std::any();
    auto* ltms = create_ltms("example", nullptr, true, true,
                              nullptr, nullptr, true, comp);
    auto* a = tms_create_node(ltms, "a", true);
    auto* b = tms_create_node(ltms, "b", true);
    auto* c = tms_create_node(ltms, "c", true);
    auto* d = tms_create_node(ltms, "d", true);
    auto* e = tms_create_node(ltms, "e", true);

    add_formula(ltms, make_or({make_not(N(a)), N(b)}));
    add_formula(ltms, make_or({make_not(N(c)), N(d)}));
    add_formula(ltms, make_or({make_not(N(c)), N(e)}));
    add_formula(ltms, make_or({make_not(N(b)), make_not(N(d)), make_not(N(e))}));
}

// ================================================================
// test_remove
// ================================================================

void test_remove() {
    auto* ltms = create_ltms("Delay", nullptr, true, true,
                              nullptr, nullptr, true, std::any(true));
    auto* a = tms_create_node(ltms, "a", true);
    auto* b = tms_create_node(ltms, "b", true);
    auto* c = tms_create_node(ltms, "c", true);

    add_formula(ltms, make_or({N(a), N(b), N(c)}));
    enable_assumption(a, NodeLabel::FALSE);
    enable_assumption(b, NodeLabel::FALSE);
    why_nodes(ltms);

    add_formula(ltms, make_or({N(a), N(b)}));
    why_nodes(ltms);
}

// ================================================================
// test_delay
// ================================================================

void test_delay() {
    auto* ltms = create_ltms("Delay", nullptr, true, true,
                              nullptr, nullptr, true, std::any(true));
    auto* a = tms_create_node(ltms, "a", true);
    auto* b = tms_create_node(ltms, "b", true);
    auto* c = tms_create_node(ltms, "c", true);

    enable_assumption(a, NodeLabel::FALSE);
    enable_assumption(b, NodeLabel::FALSE);

    add_formula(ltms, make_or({N(a), make_not(N(b))}));
    add_formula(ltms, make_or({N(b), N(c)}));

    pretty_print_clauses(ltms);
    why_nodes(ltms);

    retract_assumption(b);
    pretty_print_clauses(ltms);
    why_nodes(ltms);
}

// ================================================================
// Utility
// ================================================================

void dirty_clauses_check(LTMS* ltms) {
    int count = 0;
    walk_clauses(ltms, [&](Clause* cl) {
        if (cl->status == ClauseStatus::DIRTY) count++;
    });
    std::cout << "\n There are now " << count << " dirty clauses." << std::endl;
}

void print_statistics(LTMS* ltms) {
    std::vector<int> lengths(100, 0);
    std::cout << "\n There are " << ltms->node_counter << " propositional symbols";
    walk_clauses(ltms, [&](Clause* cl) {
        if (cl->length < 100) lengths[cl->length]++;
    });
    for (int i = 0; i < 100; i++) {
        if (lengths[i] != 0) {
            std::cout << "\n There are " << lengths[i]
                      << " prime implicates of size " << i;
        }
    }
}
